src/runtime/pythonengine.cs

```diff
@@ -302,21 +309,86 @@ public static void Shutdown()
         {
             if (initialized)
             {
-                PyScopeManager.Global.Clear();
+                // If the shutdown handlers trigger a domain unload,
+                // don't call shutdown again.
+                AppDomain.CurrentDomain.DomainUnload -= OnDomainUnload;
+
+                ExecuteShutdownHandlers();
+
                 Marshal.FreeHGlobal(_pythonHome);
                 _pythonHome = IntPtr.Zero;
                 Marshal.FreeHGlobal(_programName);
                 _programName = IntPtr.Zero;
                 Marshal.FreeHGlobal(_pythonPath);
                 _pythonPath = IntPtr.Zero;
```

Comment:
These strings are a bit weird. We're supposed not to free these ever, because they can be used even when python is not initialized. I decided that at the very least, we shouldn't free them until *after* shutting down the engine.

In python 3.7, these strings are no longer necessary: SetPythonHome and friends now copy their argument (https://github.com/python/cpython/commit/af5a895073c24637c094772b27526b94a12ec897#diff-87aed37b4704d4e1513be6378c9c7fe6) -- though that's not actually documented.

Reply:
@benoithudson  good finding, so you are cleaning up these special strings to avoid possible runtime leftovers between python re-initializations?

Reply:
In this PR I changed the order to be that we shut down python before freeing the strings, but I don't plan to change anything else.

Probably nobody will ever really notice the problem: people who set the strings will set them again next time we initialize, people who don't, won't ever set them. And in python 3.7, the problem goes away entirely. I'll create an issue, in case someone else feels motivated.